경북대 BE_김도현 2단계 - 페이지네이션

dozzzang opened the pull request:
안녕하세요 멘토님. Step2 페이지네이션 기능을 구현했습니다.

## 구현기능
- Product와 Wishlist를 조회할 때 페이지네이션 기능이 적용되도록 하였습니다.
- 이 때 페이지네이션은 기본적으로 
   페이지는 0번부터,
   7개의 항목을 하나의 Page에 보여주도록,
   Product는 상품명을 기반으로, Wishlist는 id를 기반으로 내림차순으로 구현하였으며
   `@RequestParam`을 통해 사용자가 원하는대로 page,size,sort를 조작하도록 구현하였습니다.

## 고민사항
테스트 코드를 '페이지네이션'이라는 기능에 중점을 두고 작성하고 싶어 많은 시간을 투자했습니다. 
어떤 부분을 테스트해야 효과적인지 찾아보는 과정에서 학습이 많았던 것 같습니다.
과정에서 특히 멘토님께서 1단계에서 언급해주셔서 공부했던 E2E 테스트가 생각나, E2E 테스트로 구현해보았습니다.

추가로 Step1에 대해 남겨주신 추가 피드백 사항에 대해서는 이곳 PR에 push하도록 하겠습니다. 감사합니다!!

dozzzang committed:
docs : Step2 README.md 작성

dozzzang committed:
feat : Product 페이지네이션 구현

dozzzang committed:
test : ProductPageTest E2E 통합 테스트 코드 작성
- @WebMvcTest를 통한 Controller 단위 테스트를 시도했으나, 사용하지 않는
  JwtTokenProvider,UserService 등을 모두 Mock으로 처리해야하는
  애로사항이 존재했습니다.
- 실제 DB와 연동하여 JPA의 Pageable 객체 기능을 확인하는 것이 더
  효율적이라고 느꼈습니다.

dozzzang committed:
feat : Wish 페이지네이션 구현 및 ViewController 구현

dozzzang committed:
test : Wish 페이지네이션 E2E 테스트코드 작성

mkom76 commented on src/main/java/gift/product/controller/ProductViewController.java:
```diff
@@ -21,8 +23,13 @@ public ProductViewController(ProductService productService) {
+  public String index(
+      @RequestParam(defaultValue = "0") int page,
+      @RequestParam(defaultValue = "7")int size,
+      @RequestParam(defaultValue = "name") String sort,
```
default value 설정 좋습니다👍

mkom76 commented on src/main/java/gift/wish/controller/WishViewController.java:
```diff
@@ -0,0 +1,40 @@
+  }
+
+  @GetMapping
+  public String index(
```
위시리스트 페이지까지 구현해보셨네요!👍
현재 로그인 후 페이지에 접근해도 토큰이 없다고 나오는데요, 어떻게 해결하면 좋을까요?
기본 동선은 로그인 후 버튼이 바뀌어야 하고(로그아웃, 마이페이지 버튼 추가)
로그인 이후부터는 클라이언트에서 토큰을 api에 제공해야 해요. 토큰은 일반적으로 쿠키 혹은 로컬스토리지에 저장해두었다가 api를 호출할 때에 불러와서 api header에 실어서 보내줍니다.
관련해서 질문 있으시면 연락 주세요~😊 필수 구현부는 아닌것 같아 선택적으로 진행해주셔도 될 것 같습니다!

mkom76 commented on src/test/java/gift/ProductPageTest.java:
```diff
@@ -0,0 +1,61 @@
+  }
+
+  @Test
+  void 페이지_통합_테스트() throws Exception {
```
통합테스트가 아니라 repository의 findAll 함수의 검증인 것 같아요. 단위테스트로 보여집니다.
직접 컨트롤러의 api를 호출하여 결과가 잘 나오는지 확인해보는 건 어떨까요? 아마 토큰 관련 이슈가 있을거에요. 이런 경우 테스트용 계정을 생성해두고 토큰을 가져와서 요청하는것까지 진행해보셔도 좋겠습니다!

mkom76 commented on src/main/resources/application.properties:
```diff
@@ -6,3 +6,6 @@ spring.datasource.password=
 spring.jpa.show-sql=true
+logging.level.org.hibernate.SQL=DEBUG
+logging.level.org.hibernate.type.descriptor.sql.BasicBinder=TRACE
+logging.level.org.hibernate.orm.jdbc.bind=TRACE
```
로깅용 환경설정이군요!👍
현업에서는 local, sandbox, dev, cbt, real 으로 환경이 나누어져있어 개발자들을 위한 환경에서는 로깅을 켜두고, 실제 서비스가 나가는 곳에서는 로깅을 하지 않도록 설정합니다. 참고해주세요~😊

mkom76 reviewed (changes requested):
2단계도 잘 작성해주셨네요~👍 프로젝트 완성도도 많이 높아보입니다💯
몇가지 코멘트 드렸는데요, 확인 부탁드려요!

그리고
<img width="1713" height="324" alt="image" src="https://github.com/user-attachments/assets/dfea65a6-6f09-4ab9-b778-e802a0cfeb3a" />
이 부분은 사용자들이 사용하기에 이해하기 어려울 것 같은데 카카오가 포함되지 않아 등록이 가능한 경우 true를 주는건 어떨까요?😊

dozzzang committed:
refactor(step1) : 회원가입 시 토큰발급 로직 삭제

dozzzang committed:
chore(step1) : update 메서드에 @Transactional 추가

dozzzang committed:
refactor(step2) : DB에 카카오가 포함된 product를 저장하고 있는 로직 개선

dozzzang committed:
feat : 쿠키를 이용한 로그인한 유저의 토큰 보관 기능 추가
- 로그인 시 토큰과 이메일을 쿠키에 저장
- 모든 페이지 접근 시 쿠키의 토큰을 이용하여 로그인 여부 판단
- 로그인 상태에 따른 html 버튼 동적 변경 (로그인/로그아웃 등)

dozzzang committed:
feat : user 정보 수정 폼 구현

dozzzang committed:
feat : wishlist 폼 구현 및 동작 구현

mkom76 commented on src/main/java/gift/user/controller/UserController.java:
```diff
@@ -24,12 +23,12 @@ public UserController(UserService userService) {
 
   @PostMapping("/register")
-  public ResponseEntity<RegisterResponseDto> registerUser(
+  public ResponseEntity<Void> registerUser(
```
재로그인 유도 목적일까요?😊 많은 사이트에서 사용하는 방법이죠~👍👍

dozzzang replied:
맞습니다! 피드백에서 흐름을 적어주시는 것을 보고
코드를 수정하며 어떤 흐름이 내가 경험했던 흐름이고 자연스러운 흐름인지 생각해보며 수정했습니다.

mkom76 commented on src/main/java/gift/user/controller/UserController.java:
```diff
@@ -24,12 +23,12 @@ public UserController(UserService userService) {
+    userService.registerUser(registerRequestDto);
 
-    return new ResponseEntity<RegisterResponseDto>(registerResponse, HttpStatus.CREATED);
+    return ResponseEntity.status(HttpStatus.CREATED).build();
```
정적 팩토리 메서드와 빌더 패턴을 잘 사용해주셨네요!👍👍 각 개념이 어떤 건지 숙지해두시면 좋겠어요~

mkom76 commented on src/main/java/gift/user/service/UserService.java:
```diff
@@ -72,6 +73,7 @@ public UserResponseDto findById(Long userId) {
     return UserResponseDto.from(user);
   }
 
+  @Transactional
```
Transacional을 잘 적용해주셨네요!👍
읽기에도 Transactional을 사용하기도 합니다. readOnly라는 설정을 추가해서 적용하는데요, 이를 사용하면 어떤 이점이 있을까요?😊

dozzzang replied:
@mkom76 
저번 Step1 코드리뷰에서도 같은 말씀해주신 것으로 기억하여 공부했었습니다!
``@Transactional(readOnly=true)``를 사용하면 JPA Persistent Context의 Dirty Checking으로 인한 성능 이점을 얻을 수 있습니다. Persistent Context는 엔티티 조회 시 스냅샷을 저장해두고, ``flush()`` 시점에 현재 상태와 비교하여 변경사항을 감지합니다(이것이 Dirty Checking). 변경된 내용이 있다면 UPDATE 쿼리를 생성하여 DB로 전송합니다.
readOnly = true는 JPA 세션의 Flush Mode를 Manual로 설정합니다. 즉, 개발자가 명시적으로 ``flush()`` 코드를 작성해야 하며, 그렇지 않으면 변경사항이 DB에 반영되지 않습니다. 따라서 이 메서드가 단순 조회용임을 명시하는 효과가 있고, 스냅샷을 저장하지 않아 메모리도 절약할 수 있습니다.

## 의문사항이었던 것 
### 그렇다면 그냥 annotation을 붙이지 않으면 조회할 때 snapshot을 저장하지 않으니 문제 없는 것 아닌가?
이 의문을 해결하기 위해 Lazy Loading과 Eager Loading 방식을 공부했고, Lazy Loading이 자주 사용되기 때문에 readOnly = true가 필요하다는 결론에 도달했습니다.
Lazy Loading은 연관관계 객체의 실제 데이터가 필요한 순간에 DB 쿼리를 전송하여 성능을 최적화하는 방식입니다(이전에는 Proxy 객체 상태). Lazy Loading의 근본적인 존재 이유가 객체지향적 코드 작성에 있다는 점을 고려할 때, Lazy Loading이 존재하기 때문에 @Transactional(readOnly = true)가 필요하다고 이해했습니다.
좀 더 구체적으로는 단순 조회용 메서드라도 객체지향적으로 코드를 작성하다보면 Lazy loading이 필요한 코드가 작성될 수 있는데, 그 때 Transactional annotation(메서드 자체를 하나의 영속성 컨텍스로 유지시켜야하기 때문에)이 필요하기 때문입니다.

mkom76 commented on src/main/java/gift/product/controller/ProductViewController.java:
```diff
@@ -27,7 +30,15 @@ public String index(
+      Model model,
+      HttpServletRequest request) {
+
+    String token = getCookieValue(request, "token");
```
쿠키 추가 좋습니다!👍👍 개발자모드에서도 잘 보이네요~

mkom76 reviewed (approved):
스프링 백엔드 완성도도 높고 선택사항인 프론트까지 잘 구현해주셨네요!👍👍💯
고생 많으셨어요. 3단계도 화이팅입니다!💪💪💪
